test/test_ensemble_builder/test_ensemble_builder.py

```diff
@@ -0,0 +1,554 @@
+    -------
+    * Should filter out runs with no loss
+    """
+    bad_runs = [make_run(loss=None) for _ in range(5)]
+    dummy_run = make_run(dummy=True, loss=2)
+    good_run = make_run(loss=1)
+
+    runs = bad_runs + [good_run]
+
+    candidates, discarded = builder.candidate_selection(runs, dummy_run)
+
+    assert len(candidates) == 1
+    assert len(discarded) == len(bad_runs)
+    assert candidates[0].loss == 1
+
+
+def test_candidates_filters_out_better_than_dummy(
+    builder: EnsembleBuilder,
+    make_run: Callable[..., Run],
+) -> None:
+    """
+    Expects
+    -------
+    * Should filter out runs worse than dummy
+    """
+    bad_runs = [make_run(loss=1) for _ in range(2)]
+    dummy_run = make_run(dummy=True, loss=0)
+    good_runs = [make_run(loss=-1) for _ in range(3)]
+
+    runs = bad_runs + good_runs
+
+    candidates, discarded = builder.candidate_selection(
+        runs, dummy_run, better_than_dummy=True
+    )
+
+    assert set(candidates)
+
+    assert len(candidates) == 3
+    assert all(run.loss < dummy_run.loss for run in candidates)
+
+    assert len(discarded) == 2
+    assert all(run.loss >= dummy_run.loss for run in discarded)
+
+
+def test_candidates_uses_dummy_if_no_candidates_better(
+    builder: EnsembleBuilder,
+    make_run: Callable[..., Run],
+) -> None:
+    """
+    Expects
+    -------
+    * If no run is better than a dummy run, the candidates will then consist
+      of the dummy runs.
+    """
+    runs = [make_run(loss=10) for _ in range(10)]
+    dummies = [make_run(dummy=True, loss=0) for _ in range(2)]
+
+    candidates, discarded = builder.candidate_selection(
+        runs,
+        dummies,
+        better_than_dummy=True,
+    )
+
+    assert len(candidates) == 2
+    assert all(run.is_dummy() for run in candidates)
+
+
+@parametrize("nbest", [0, 1, 5, 1000])
+def test_candidates_nbest_int(
+    builder: EnsembleBuilder,
+    make_run: Callable[..., Run],
+    nbest: int,
+) -> None:
+    """
+    Expects
+    -------
+    * Should only select the nbest candidates
+    * They should be ordered by loss
+    """
+    n = 10
+    expected = int(bound(nbest, bounds=(1, n)))
+
+    dummy = make_run(dummy=True)
+    runs = [make_run(loss=i) for i in range(n)]
+    candidates, discarded = builder.candidate_selection(runs, dummy, nbest=nbest)
+
+    assert len(candidates) == expected
+
+    if len(candidates) > 1:
+        assert all(a.loss <= b.loss for a, b in pairs(candidates))
+
+    if any(discarded):
+        worst_candidate = candidates[-1]
+        assert all(worst_candidate.loss <= d.loss for d in discarded)
+
+
+@parametrize("nbest", [0.0, 0.25, 0.5, 1.0])
+def test_candidates_nbest_float(
+    builder: EnsembleBuilder,
+    make_run: Callable[..., Run],
+    nbest: float,
+) -> None:
+    """
+    Expects
+    -------
+    * Should select nbest percentage of candidates
+    * They should be ordered by loss
+    """
+    n = 10
+    expected = int(bound(nbest * n, bounds=(1, n)))
```

Comment:
Should there also be a max(1, ...) around this to ensure we expect at least one model?

Reply:
The bound function bounds thing between the bounds.

i.e. `bound(x, bounds=(1, 30))` means we will always get a number between `1` and `30`.

This is the definition of `bound`, it's written in `util/functional.py` as I thought it's something that will probably be re-used

```python
def bound(val: float, bounds: tuple[float, float]) -> float:
    """Bounds a value between a low and high

    .. code:: python

        x = bound(14, low=0, high=13.1)
        # x == 13.1

    Parameters
    ----------
    val : float
        The value to bound

    bounds: tuple[foat, float]
        The bounds to bound the value between (low, high)

    Returns
    -------
    float
        The bounded value
    """
    return max(bounds[0], min(val, bounds[1]))
```

Reply:
Sorry, I missed `bound` inside the int.